include/alpaka/core/CallbackThread.hpp

```diff
@@ -0,0 +1,77 @@
+/* Copyright 2022 Antonio Di Pilato
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace alpaka::core
+{
+    class CallbackThread
+    {
+        using Task = std::packaged_task<void()>;
+
+    public:
+        ~CallbackThread()
+        {
+            {
+                std::unique_lock<std::mutex> lock{m_mutex};
+                m_stop = true;
+            }
+            m_cond.notify_one();
+            if(m_thread.joinable())
+                m_thread.join();
+        }
+        auto submit(Task&& newTask) -> std::future<void>
+        {
+            auto f = newTask.get_future();
+            {
+                std::unique_lock<std::mutex> lock{m_mutex};
+                m_tasks.emplace(std::move(newTask));
+                if(!m_thread.joinable())
+                {
+                    m_thread = std::thread(
+                        [this]
+                        {
+                            Task task;
+                            while(true)
+                            {
+                                {
+                                    std::unique_lock<std::mutex> lock{m_mutex};
+                                    m_cond.wait(lock, [this] { return m_stop || !m_tasks.empty(); });
+
+                                    if(m_stop && m_tasks.empty())
+                                        break;
+
+                                    task = std::move(m_tasks.front());
+                                    m_tasks.pop();
+                                }
+
+                                task();
+                            }
+                        });
+                }
+            }
+            m_cond.notify_one();
+            return f;
+        }
+
+    private:
+        std::thread m_thread;
+        std::condition_variable m_cond;
+        std::mutex m_mutex;
+        bool m_stop{false};
```

Comment:
If you make `m_stop` an `std::atomic_bool` or an `std::atomic_flag` you can avoid the lock in the destructor.
Not sure it's worth the change, though.

Reply:
It is worthwhile. Using an atomic instead of a lock gives us more exception safety in the destructor. (Not that we could do anything sensible when one of these standard library calls fails, but it is nice to be more "legal".)

Reply:
However, changing it to an atomic makes the destructor faster, but pessimizes all loads to `m_stop` inside the thread loop. It is also an unusual design to me.

Reply:
> Using an atomic instead of a lock gives us more exception safety in the destructor.

How so? The ctor and dtor of `unique_lock` are `noexcept`.

Reply:
It seems I slipped one row below. Sorry for the fuss!